test/explicit-resource-management/main.test.ts

```diff
@@ -0,0 +1,309 @@
+
+import { describe, it } from 'mocha';
+import { GridFSBucket, MongoClient } from 'mongodb/lib/beta';
+import { Readable } from 'stream';
+import { pipeline } from 'stream/promises';
+import { expect } from 'chai';
+import { setTimeout } from 'timers/promises';
+import { createReadStream } from 'fs';
+import { join } from 'path';
+
+async function setUpCollection(client: MongoClient) {
+  const collection = client.db('foo').collection<{ name: string }>('bar');
+  const documents: Array<{ name: string }> = Array.from({ length: 5 }).map(i => ({
+    name: String(i)
+  }));
+  await collection.insertMany(documents)
+  return collection;
+}
+
+describe('explicit resource management feature integration tests', function () {
+  describe('MongoClient', function () {
+    it('does not crash or error when used with await-using syntax', async function () {
+      await using client = new MongoClient(process.env.MONGODB_URI!);
+      await client.connect();
+    })
+
+    it('always cleans up the client, regardless of thrown errors', async function () {
```

Comment:
for the cases that say "cleans up X", can we make the assertion somehow relate to the clean up (either assert that the resource is closed or that the relevant method was called with the expected arguments where applicable?) or, if not,  can we align the test case description with what the test actually guarantees?

Reply:
I'd prefer not to.  I think "cleans up x resource" is accurate.  I also don't want to introduce any sinon spying into this test suite - we are already covered with sinon assertions asserting we call the correct disposal method Symbol.asyncDispose is called in the manual integration tests.

Reply:
Definitely a good addition, although it doesn't help us here.  We can't check `client.closed` because after we leave scope there is no client 🙂

Reply:
```js
let autoCleanedUpClient;
const error = await (async () => {
  await using client = new MongoClient(process.env.MONGODB_URI!);
  autoCleanedUpClient = client;
  // ...
expect(autoCleanedUpClient).to.have.nested.property('s.hasBeenClosed', true); 
```

Seems like you can save what you need to check outside of the scope. And we can use what we already have internally instead of a `closed` prop 

(although I don't disagree, that `x.closed` is nice, and maybe a pattern we should make sure is shared on everything in design close work)

Reply:
Okay, I feel that perhaps I haven't adequately communicated this.  Spying is an option, so is saving a reference to the client, or other things like overriding the dispose method with a custom implementation, etc.

I feel that any testing of whether or not `Symbol.asyncDispose` is called or not is just testing the TS polyfill, and that since we already have tests that use sinon to assert that the correct cleanup methods are called when `Symbol.asyncDispose` is called, asserting on client state / spying in the feature smoke tests isn't necessary.

Reply:
My recommendation is based on two testing best practices:
1) test prose should accurately describe what the author intended to safeguard by writing the test
2) tests should _always_ fail if what they say *should* happen does not happen

Both of these speak to the "DAMP not DRY" idea of testing ([here](https://abseil.io/resources/swe-book/html/ch12.html) is a good read on the subject of writing clear tests with that principle in context).

So, if a test says, e.g., "always cleans up the client", then that test needs to fail if the client is not, in fact, cleaned up. Right now, this test won't fail if the `asyncDispose` becomes a no-op. That behavior would be surprising to any future developer looking at this test case. This can be fixed either by adjusting the assertions or the description. Either is acceptable to me personally, though the consensus among the three reviewers here seems to be in favor of assertions on the cleanup.

Reply:
I added sinon assertions to these tests.